python/src/iceberg/transforms.py

```diff
@@ -0,0 +1,520 @@
+            lambda t: t == DoubleType,
+        ),
+    }
+
+    @staticmethod
+    def _decimal_to_bytes(value: Decimal):
+        unscaled_value = Transform._unscale_decimal(value)
+        number_of_bytes = int(math.ceil(unscaled_value.bit_length() / 8))
+        return unscaled_value.to_bytes(length=number_of_bytes, byteorder="big")
+
+    def __init__(self, transform_type: Type, num_buckets: int):
+        if (
+            transform_type not in Bucket._FUNCTIONS_MAP
+            and not isinstance(transform_type, FixedType)
+            and not isinstance(transform_type, DecimalType)
+        ):
+            raise ValueError(f"Cannot bucket by type: {transform_type}")
+
+        super().__init__(
+            f"bucket[{num_buckets}",
+            f"transforms.bucket(transform_type={repr(transform_type)}, num_buckets={num_buckets})",
+        )
+        self._type = transform_type
+        self._num_buckets = num_buckets
+
+    @property
+    def num_buckets(self) -> int:
+        return self._num_buckets
+
+    def apply(self, value):
+        if value is None:
+            return None
+
+        if isinstance(self._type, FixedType):
+            return (
+                Bucket._MURMUR3.hash(value) & Bucket._MAX_32_BITS_INT
+            ) % self._num_buckets
+        elif isinstance(self._type, DecimalType):
+            return (
+                Bucket._MURMUR3.hash(Bucket._decimal_to_bytes(value))
+                & Bucket._MAX_32_BITS_INT
+            ) % self._num_buckets
+
+        return (
+            Bucket._FUNCTIONS_MAP[self._type][0](value) & Bucket._MAX_32_BITS_INT
+        ) % self._num_buckets
+
+    def can_transform(self, target: Type) -> bool:
+        if isinstance(self._type, FixedType):
+            return target == BinaryType or isinstance(target, FixedType)
+        elif isinstance(self._type, DecimalType):
+            return isinstance(target, DecimalType)
+
+        return Bucket._FUNCTIONS_MAP[self._type][1](target)
+
+    def get_result_type(self, source: Type):
+        return IntegerType
+
+    def to_human_string(self, value):
+        return str(value)
+
+
+class Time(Transform):
+    """
+    Time class is for both Date transforms and Timestamp transforms.
+    """
+
+    _TIME_ORDER = datetime(year=3, month=2, day=1, hour=0)
+    _VALID_TIME_GRANULARITY = {
+        DateType: {"year", "month", "day"},
+        TimestampType: {"year", "month", "day", "hour"},
+        TimestamptzType: {"year", "month", "day", "hour"},
+    }
+    _DIFF_MAP = {
+        "year": lambda t1, t2: (t1.year - t2.year)
+        - (
+            1
+            if t1.month < t2.month or (t1.month == t2.month and t1.day < t2.day)
+            else 0
+        ),
+        "month": lambda t1, t2: (t1.year - t2.year) * 12
+        + (t1.month - t2.month)
+        - (1 if t1.day < t2.day else 0),
+        "day": lambda t1, t2: (t1 - t2).days,
+        "hour": lambda t1, t2: int((t1 - t2).total_seconds() / 3600),
```

Comment:
This lambda format is very hard to read. Can you write these as functions and place them in the map instead of lambdas?

Reply:
Yep, move them to transform_util as module methods